### PR TITLE
Room service progress

### PR DESCRIPTION
**Pull request by mariellenolasco**
# housingxyz :: pull-request

## Summary

- Added more functionality to complex controller. Fixed some errors in the service bus (i.e. the not implemented exception in the ExecuteAsync). 
## reviewer

- @escalonn

## contributing

this project is maintained under the terms of the [Contribution Guidelines][contribution-guidelines-url].

[contribution-guidelines-url]: https://github.com/housingxyz/housingxyz/blob/master/.github/CONTRIBUTING.md 'CONTRIBUTION GUIDELINES'

### CODE REVIEW

File: room/src/Revature.Room.Api/Controllers/ComplexController.cs [hunk 1]
```diff
@@ -68,5 +70,167 @@ public ComplexController(IRepository repository, ILogger logger)
...
+      } catch (ArgumentException ex)
+      {
+        _logger.LogInformation(ex.Message);
+        return BadRequest();
+      }
+
+    }
+
+    /// <summary>
+    /// Update a room given the provided their id: Guid
```
Comment (g-b-ang) [inline room/src/Revature.Room.Api/Controllers/ComplexController.cs]: Could make documentation specify that only lease gets updated

File: room/src/Revature.Room.Api/Controllers/ComplexController.cs [hunk 2]
```diff
@@ -68,5 +70,167 @@ public ComplexController(IRepository repository, ILogger logger)
...
+          RoomType = room.RoomType
+        };
+        createdRoom.SetLease(room.LeaseStart, room.LeaseEnd);
+        await _repository.CreateRoomAsync(createdRoom);
+        await _repository.SaveAsync();
+
+        _logger.LogInformation("Success. Room has been added");
+
+        return CreatedAtRoute("GetRoom", new { RoomID = createdRoom.RoomId }, createdRoom);
+      } catch (ArgumentException ex)
```
Comment (g-b-ang) [inline room/src/Revature.Room.Api/Controllers/ComplexController.cs]: change to

}
catch (ArgumentException ex)
{

to make formatting consistent

File: room/src/Revature.Room.Api/Controllers/ComplexController.cs [hunk 3]
```diff
@@ -68,5 +70,167 @@ public ComplexController(IRepository repository, ILogger logger)
...
+      {
+        _logger.LogInformation("Deleting rooms from complex");
+        Lib.Room co = new Lib.Room();
+        co.ComplexId = id;
+
+        var listOfGuid = await _repository.DeleteComplexRoomAsync(co.ComplexId);
+        await _repository.SaveAsync();
+
+        await _busSender.SendDeleteComplexMessage(listOfGuid);
+        _logger.LogInformation("Success! Rooms have been deleted");
```
Comment (g-b-ang) [inline room/src/Revature.Room.Api/Controllers/ComplexController.cs]: the log for "rooms have been deleted" should go after repo.save() and then an additional log after the busSender.sendDeleteMsg() can be a more busSender specific log

File: room/src/Revature.Room.Api/Services/ServiceBusConsumer.cs [hunk 4]
```diff
@@ -141,7 +152,8 @@ public async Task CloseQueueAsync()
     /// <exception cref="NotImplementedException">Inherited but not utilized</exception>
     protected override Task ExecuteAsync(CancellationToken stoppingToken)
     {
-      throw new NotImplementedException();
+      RegisterOnMessageHandlerAndReceiveMessages();
```
Comment (g-b-ang) [inline room/src/Revature.Room.Api/Services/ServiceBusConsumer.cs]: can this be logged?

File: room/src/Revature.Room.DataAccess/Repository.cs [hunk 5]
```diff
@@ -145,21 +162,42 @@ public async Task SaveAsync()
...
       return await _context.Room
-        .Where(r => r.Gender.Type.ToUpper() == gender.ToUpper() && endDate < r.LeaseEnd && r.NumberOfOccupants < r.NumberOfBeds)
-        .Select(r => r.RoomId).ToListAsync();
+        .Where(r => (r.Gender == null || r.Gender.Type.ToUpper() == gender.ToUpper()) && endDate < r.LeaseEnd && r.NumberOfOccupants < r.NumberOfBeds)
+        .Select(r => new Tuple<Guid, int>( r.RoomId, r.NumberOfBeds))
+        .ToListAsync();
     }
+
     /// <summary>
     /// Method that updates the room occupants when a tenant is assigned a room
```
Comment (g-b-ang) [inline room/src/Revature.Room.DataAccess/Repository.cs]: Also updates the room's gender if previously null

File: room/src/Revature.Room.DataAccess/Repository.cs [hunk 6]
```diff
@@ -145,21 +162,42 @@ public async Task SaveAsync()
...
-      Entities.Room roomToUpdate = await _context.Room.FirstAsync(r => r.RoomId == roomId);
+      Entities.Room roomToUpdate = await _context.Room.Where(r => r.RoomId == roomId).Include(r => r.Gender).FirstAsync();
       roomToUpdate.NumberOfOccupants++;
+      if(roomToUpdate.Gender == null)
+      {
+        roomToUpdate.Gender = await _context.Gender.FirstAsync(g => g.Type.ToUpper() == tenantGender.ToUpper());
+      }
+    }
+    /// <summary>
+    /// Method that updates occupants when an occupant vacates a room
```
Comment (g-b-ang) [inline room/src/Revature.Room.DataAccess/Repository.cs]: Also updates gender to null if room is left empty

File: room/src/Revature.Room.Lib/IRepository.cs [hunk 7]
```diff
@@ -72,13 +72,27 @@ public interface IRepository
...
+    /// <summary>
+    /// Method that updates occupants when an occupant vacates a room
+    /// </summary>
+    /// <param name="roomId"></param>
+    /// <returns></returns>
+    /// <exception cref="InvalidOperationException">Thrown when room isn't found</exception>
+    public Task SubtractRoomOccupantsAsync(Guid roomId);
 
+    /// <summary>
+    /// Deletes a complex and deletes all rooms that is connected to that complex
```
Comment (g-b-ang) [inline room/src/Revature.Room.Lib/IRepository.cs]: Technically does not delete a complex, just all rooms connected to it

File: room/src/Revature.Room.Api/Services/ServiceBusConsumer.cs (hunk 4 above)
Comment (mariellenolasco) [reply room/src/Revature.Room.Api/Services/ServiceBusConsumer.cs]: I don't know

Comment (mariellenolasco) [reply room/src/Revature.Room.Api/Services/ServiceBusConsumer.cs]: Should we log it?

File: room/.dockerignore [hunk 8]
```diff
@@ -22,9 +23,6 @@
 **/bin/
 **/obj/
 
-**/appsettings.*
-!**/appsettings.json
-
```
Comment (escalonn) [inline room/.dockerignore]: remove the diff to this file; remember we took a different approach, with a dockerup.localhost.yaml

File: room/src/Revature.Room.Api/Controllers/ComplexController.cs [hunk 9]
```diff
@@ -68,5 +73,180 @@ public ComplexController(IRepository repository, ILogger logger)
...
+
+        var result = await _repository.ReadRoomAsync(roomId);
+
+        _logger.LogInformation("Success");
+
+        return Ok(result);
+      }
+      catch (InvalidOperationException ex)
+      {
+        _logger.LogError(ex.Message);
```
Comment (escalonn) [inline room/src/Revature.Room.Api/Controllers/ComplexController.cs]: don't use `ex.Message` in logs; all the log methods already have overloads to log the whole exception object with its stack trace

File: room/src/Revature.Room.Api/Controllers/ComplexController.cs [hunk 10]
```diff
@@ -68,5 +73,180 @@ public ComplexController(IRepository repository, ILogger logger)
...
+      }
+    }
+
+    /// <summary>
+    /// Creates a room based on the complex's needs
+    /// </summary>
+    /// <param name="room"></param>
+    /// <returns></returns>
+    [HttpPost]
+    [ProducesResponseType(StatusCodes.Status201Created)]
```
Comment (escalonn) [inline room/src/Revature.Room.Api/Controllers/ComplexController.cs]: pass the 2nd parameter to this attribute, `typeof(Revature.Room.Lib.Room)` i think. this tells swagger what the response body will look like, which it wants to know.

File: room/src/Revature.Room.Api/Controllers/ComplexController.cs [hunk 11]
```diff
@@ -68,5 +73,180 @@ public ComplexController(IRepository repository, ILogger logger)
...
+    /// </summary>
+    /// <param name="id"></param>
+    /// <param name="Lroom"></param>
+    /// <returns></returns>
+    /// <remarks>Update room functionality of complex service</remarks>
+    [HttpPut("{id}")]
+    [ProducesResponseType(StatusCodes.Status204NoContent)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    public async Task<IActionResult> PutRoomAsync(Guid id, [FromBody] Lib.Room Lroom)
```
Comment (escalonn) [inline room/src/Revature.Room.Api/Controllers/ComplexController.cs]: naming convention

File: room/src/Revature.Room.Api/Services/ServiceBusConsumer.cs [hunk 12]
```diff
@@ -19,7 +19,13 @@ namespace ServiceBusMessaging
   /// </summary>
   public class ServiceBusConsumer : BackgroundService, IServiceBusConsumer
   {
+
+    //_queueClient will be used for receiving from the Complex
     private readonly QueueClient _queueClient;
+
+    //_queueClientTwo will be used for receiving from the Tenant
+    private readonly QueueClient _queueClientTwo;
```
Comment (escalonn) [inline room/src/Revature.Room.Api/Services/ServiceBusConsumer.cs]: the point of service bus is looser coupling to the other services, so the primary fact about them is you listen/send messages of a given type, not the source/destination service. the names should be more descriptive on this basis

File: room/src/Revature.Room.Api/Services/ServiceBusConsumer.cs (hunk 4 above)
Comment (escalonn) [reply room/src/Revature.Room.Api/Services/ServiceBusConsumer.cs]: what's so different about logging this? "Registered Service Bus consumer."

File: room/src/Revature.Room.Api/Services/ServiceBusSender.cs [hunk 13]
```diff
@@ -25,17 +27,35 @@ public class ServiceBusSender
...
+    /// <param name="roomToSend"></param>
+    /// <returns></returns>
+    public async Task SendDeleteComplexMessage(List<Guid> roomToSend)
+    {
+      string data = JsonSerializer.Serialize(roomToSend);
+
+      Message message = new Message(Encoding.UTF8.GetBytes(data));
+
+      _logger.LogInformation("ServiceBus sending delete guid message: ", data);
+      await _queueClient.SendAsync(message);
```
Comment (escalonn) [inline room/src/Revature.Room.Api/Services/ServiceBusSender.cs]: there is some common code here which you can extract to helper method(s)

File: room/src/Revature.Room.Api/Startup.cs [hunk 14]
```diff
@@ -55,6 +54,9 @@ public void ConfigureServices(IServiceCollection services)
       services.AddScoped<IMapper, DBMapper>();
       services.AddHostedService<ServiceBusConsumer>();
 
+      //Added this for sender because now we are sending things 11-22-19
```
Comment (escalonn) [inline room/src/Revature.Room.Api/Startup.cs]: why is this comment here

File: room/.dockerignore (hunk 8 above)
Comment (mariellenolasco) [reply room/.dockerignore]: okayyyyy

**Comment by escalonn** [comment]
get another approve

File: room/src/Revature.Room.Api/Services/IServiceBusSender.cs [hunk 15]
```diff
@@ -0,0 +1,11 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
```
Comment (markcmoore) [inline room/src/Revature.Room.Api/Services/IServiceBusSender.cs]: commentary

**Comment by escalonn** [comment]
gotdangit marielle i said get another approve first

**Pull request merged**